File: backend/tools/system_tools/reload_tools.py
import logging

logger = logging.getLogger(__name__)

TOOL_DEF = {
    "name": "reload_tools",
    "description": "Reload tools from backend/tools (including subfolders). Call this after creating or editing a tool module so Akira can use the new or updated tool without restarting the server.",
    "input_schema": {"type": "object", "properties": {}},
    "default_enabled": True,
}


def call_tool(tool_input: dict, context=None):
    if context is None or not hasattr(context, "_reload_tools"):
        return 500, {"success": False, "error": "Reload not available (no context)."}
    try:
        context._reload_tools()
        return 200, {"success": True, "message": "Tools reloaded. New tools under backend/tools are now available."}
    except Exception as e:
        logger.exception("Failed to reload tools")
        return 500, {"success": False, "error": str(e)}

File: akira-desktop/electron/tools/system/reload-tools.js
/**
 * Reload Tools Tool
 * reload_tools - Reload tools from the tools directory
 */

// Reload callback - will be set by the main process
let reloadCallback = null;

function setReloadCallback(callback) {
  reloadCallback = callback;
}

const definitions = [
  {
    name: 'reload_tools',
    description: 'Reload tools from the tools directory. Call this after creating or editing a tool module so Akira can use the new or updated tool without restarting.',
    input_schema: {
      type: 'object',
      properties: {},
    },
  },
];

const handlers = {
  async reload_tools(input) {
    if (!reloadCallback) {
      return {
        success: false,
        error: 'Reload not available (no callback configured).',
      };
    }

    try {
      await reloadCallback();
      return {
        success: true,
        message: 'Tools reloaded. New tools are now available.',
      };
    } catch (error) {
      return {
        success: false,
        error: error.message || String(error),
      };
    }
  },
};

module.exports = { definitions, handlers, setReloadCallback };
